ui/src/services/csrfTokenApi.ts

```diff
@@ -0,0 +1,125 @@
+
+export type CsrfToken = {
+    token: string | null;
+    expiration: Date | null;
+    client: string;
+    issued: Date | null;
+}
+
+export type CsrfTokenResponse = {
+    token: string;
+    expiration: Date;
+}
+
+export class CsrfTokenApi extends Api {
+    public csrfToken: CsrfToken = {
+        token: null,
+        expiration: null,
+        client: randomId(),
+        issued: null,
+    };
+    public csrfSupportEnabled = true;
+    private refreshTimeout: ReturnType<typeof setTimeout> | null = null;
+    private ongoingRefresh: Promise<void> | null = null;
+
+    public constructor(apiConfig: PrefectConfig, instanceSetupHook: AxiosInstanceSetupHook | null = null) {
+        super(apiConfig, instanceSetupHook)
+        this.startBackgroundTokenRefresh()
+    }
+
+    public async addCsrfHeaders(config: InternalAxiosRequestConfig) {
+        if (this.csrfSupportEnabled) {
+            await this.refreshCsrfToken()
+            config.headers = config.headers || {}
+            config.headers['Prefect-Csrf-Token'] = this.csrfToken.token
+            config.headers['Prefect-Csrf-Client'] = this.csrfToken.client
+        }
+    }
+
+    private async refreshCsrfToken(force: boolean = false) {
+        if (!force && !this.shouldRefreshToken()) {
+            return this.ongoingRefresh ?? Promise.resolve();
+        }
+
+        if (this.ongoingRefresh) {
+            return this.ongoingRefresh;
+        }
+
+        const refresh = async () => {
+            try {
+                const response = await this.get<CsrfTokenResponse>(`/csrf-token?client=${this.csrfToken.client}`);
+                const now = new Date();
+                this.csrfToken.token = response.data.token;
+                this.csrfToken.expiration = response.data.expiration;
+                this.csrfToken.issued = now;
+                this.ongoingRefresh = null;
+            } catch (error) {
+                this.ongoingRefresh = null;
+                if (isAxiosError(error)) {
+                    const status = error.response?.status;
+                    const unconfiguredServer = status === 422 && error.response?.data.detail.includes('CSRF protection is disabled.');
+
+                    if (unconfiguredServer) {
+                        this.disableCsrfSupport()
+                    } else {
+                        console.error('Failed to refresh CSRF token:', error)
+                        showToast('Failed to refresh CSRF token')
+                        throw new Error('Failed to refresh CSRF token')
+                    }
+                }
+            }
+        };
+
+        this.ongoingRefresh = refresh();
+        return this.ongoingRefresh;
+    }
+
+    private shouldRefreshToken(): boolean {
+        const now = new Date()
+        return !!(this.csrfSupportEnabled && (!this.csrfToken.token || (this.csrfToken.expiration && now > this.csrfToken.expiration)))
+    }
```

Comment:
Is this the same? I was having a hard time with the amount of logical operations and parens. 
```suggestion
    private shouldRefreshToken(): boolean {
        if(!this.csrfSupportEnabled) {
          return false
        }
        
        if(!this.csrfToken.token || !this.csrfToken.expiration) {
          return true
        }
        
        const now = new Date()
        
        return now > this.csrfToken.expiration
    }
```

Reply:
Looks the same to my eyes, and cleaner, thanks !